### PR TITLE
Changes and fixes for time series classification

### LINKED ISSUE #1681: Changes and fixes for time series classification problems
Comment with reproducer thanks to @bchen1116 :

https://github.com/alteryx/evalml/pull/1666#pullrequestreview-566400601

### PR DESCRIPTION
**Pull request by ParthivNaresh**
Fixes #1681 #3095 #3125 

I'd recommend starting with `evalml/utils/gen_utils.py`

### CODE REVIEW

File: evalml/automl/automl_search.py [hunk 1]
```diff
@@ -150,21 +150,6 @@ def search(
     y_train = infer_feature_types(y_train)
     problem_type = handle_problem_types(problem_type)
 
-    datetime_column = None
```
Comment (ParthivNaresh) [inline evalml/automl/automl_search.py:153]: Removing this because `DefaultDataChecks` are run anyways which includes `TimeSeriesParametersDataCheck`

File: evalml/automl/automl_search.py [hunk 2]
```diff
@@ -795,8 +769,6 @@ def _validate_problem_configuration(self, problem_configuration=None):
             is_valid, msg = contains_all_ts_parameters(problem_configuration)
             if not is_valid:
                 raise ValueError(msg)
-            if problem_configuration["date_index"] is None:
```
Comment (ParthivNaresh) [inline evalml/automl/automl_search.py:798]: `contains_all_ts_parameters` covers this case

File: evalml/automl/engine/engine_base.py [hunk 3]
```diff
@@ -140,6 +140,8 @@ def train_pipeline(pipeline, X, y, automl_config, schema=True):
         cv_pipeline.problem_type,
         X_threshold_tuning,
         y_threshold_tuning,
+        X,
+        y,
```
Comment (ParthivNaresh) [inline evalml/automl/engine/engine_base.py:144]: Passed in so `tune_binary_threshold` has access to the original `X` and `y`.

File: evalml/automl/utils.py [hunk 4]
```diff
@@ -123,7 +132,14 @@ def tune_binary_threshold(
...
         pipeline.threshold = 0.5
         if X_threshold_tuning is not None:
-            y_predict_proba = pipeline.predict_proba(X_threshold_tuning)
+            if is_time_series(problem_type):
+                X_threshold_tuning = X_threshold_tuning.iloc[
+                    : pipeline.forecast_horizon
+                ]
+                y_threshold_tuning = y_threshold_tuning.iloc[
+                    : pipeline.forecast_horizon
+                ]
```
Comment (ParthivNaresh) [inline evalml/automl/utils.py:141]: `predict_proba` is limited to `forecast_horizon` observations

File: docs/source/release_notes.rst [hunk 5]
```diff
@@ -7,11 +7,14 @@ Release Notes
         * Added ``TimeSeriesParametersDataCheck`` to verify the time series parameters are valid given the number of splits in cross validation :pr:`3111`
     * Fixes
         * Default parameters for ``RFRegressorSelectFromModel`` and ``RFClassifierSelectFromModel`` has been fixed to avoid selecting all features :pr:`3110`
+        * Fixed error caused when tuning threshold for time series binary classification :pr: `3135`
     * Changes
         * Removed reliance on a datetime index for ``ARIMARegressor`` and ``ProphetRegressor`` :pr:`3104`
         * Included target leakage check when fitting ``ARIMARegressor`` to account for the lack of ``TimeSeriesFeaturizer`` in ``ARIMARegressor`` based pipelines :pr:`3104`
         * Cleaned up and refactored ``InvalidTargetDataCheck`` implementation and docstring :pr:`3122`
         * Removed indices information from the output of ``HighlyNullDataCheck``'s ``validate()`` method :pr:`3092`
+        * ``TimeSeriesParametersDataCheck`` was expanded to consider inadequate class representation during data splitting for time series classification :pr:`3135`
```
Comment (angela97lin) [inline docs/source/release_notes.rst:16]: This is a lot for one PR! Is there a way to split this up into multiple PRs--it'd be easier to review :)

Comment (ParthivNaresh) [reply docs/source/release_notes.rst:16]: Hmm I think I can split adding `TimeSeriesParametersDataCheck` to `DefaultDataChecks` into a different PR!

**Comment by freddyaboulton** [comment]
@ParthivNaresh FYI you have to add "Fixes" before each issue for github to automatically close them. So "Fixes #1681, Fixes #3095, ..." I saw the same thing happened with #3104

File: evalml/utils/gen_utils.py [hunk 6]
```diff
@@ -556,23 +564,67 @@ def are_ts_parameters_valid_for_split(
...
     window_size = gap + max_delay + forecast_horizon
     msg = ""
+
+    is_split_less_than_window = False
+    is_split_one_class = False
+    invalid_splits = []
+
     if split_size <= window_size:
+        is_split_less_than_window = True
+    if y is not None and handle_problem_types(problem_type) in [
```
Comment (freddyaboulton) [inline evalml/utils/gen_utils.py:590]: @ParthivNaresh Quick question: at OH we discussed that this check only needed to apply to the first fold. Why the change here? Can we simplify this by only checking the first fold?

File: evalml/data_checks/ts_parameters_data_check.py [hunk 7]
```diff
@@ -1,64 +1,121 @@
...
 from evalml.utils.gen_utils import (
-    are_ts_parameters_valid_for_split,
     contains_all_ts_parameters,
+    is_ts_valid_for_split,
 )
 
 
 class TimeSeriesParametersDataCheck(DataCheck):
-    """Checks whether the time series parameters are compatible with data splitting.
+    """Checks whether the time series parameters and target data are compatible with data splitting.
```
Comment (freddyaboulton) [inline evalml/data_checks/ts_parameters_data_check.py:11]: I argue we keep `TimeSeriesParametersDataCheck` the same, along with `are_ts_parameters_valid_for_split`, and add a new data check/util method for checking the class imbalance for time series classification.

My reasoning is that although the classification imbalance check and split size vs window size check are related, they are conceptually checking different things and that maps nicer to independent data checks and methods. Plus, in the future, if we decide we need to augment/improve our implementation for checking imbalanced classes for time series we can do that independently of the implementation for the split-size vs window size check.

**Review by freddyaboulton — COMMENTED** [review]
@ParthivNaresh Looks good! I think we should have separate data checks and util methods for the class imbalance data check.

File: evalml/automl/utils.py (hunk 4 above)
Comment (freddyaboulton) [reply evalml/automl/utils.py:141]: I think we can use `predict_proba_in_sample` here?
